rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
+            req.respond(
+                &self.handler,
+                &exchange.accessor()?,
+                &metadata.node(),
+                &mut wb,
+            )
+            .await?
+        };
+
+        exchange.send(OpCode::WriteResponse, wb.as_slice()).await?;
+
+        Ok(more_chunks)
+    }
+
+    async fn invoke(
+        &self,
+        exchange: &mut Exchange<'_>,
+        timeout_instant: Option<Duration>,
+    ) -> Result<(), Error> {
+        let req = InvReq::from_tlv(&get_root_node_struct(exchange.rx()?.payload())?)?;
+        debug!("IM: Invoke request: {:?}", req);
+
+        let timed = req.timed_request.unwrap_or(false);
+
+        if self.timed_out(exchange, timeout_instant, timed).await? {
+            return Ok(());
+        }
+
+        let Some(mut tx) = self.tx_buffer(exchange).await? else {
+            return Ok(());
+        };
+
+        let mut wb = WriteBuf::new(&mut tx);
+
+        let metadata = self.handler.lock().await;
+
+        let req = InvReq::from_tlv(&get_root_node_struct(exchange.rx()?.payload())?)?;
+
+        let awaits = metadata
+            .node()
+            .invoke(&req, &exchange.accessor()?)
+            .any(|item| {
+                item.map(|(cmd, _)| self.handler.invoke_awaits(&cmd))
+                    .unwrap_or(false)
+            });
+
+        if awaits {
+            let Some(rx) = self.rx_buffer(exchange).await? else {
+                return Ok(());
+            };
+
+            let req = InvReq::from_tlv(&get_root_node_struct(&rx)?)?;
+
+            req.respond(&self.handler, exchange, &metadata.node(), &mut wb)
+                .await?;
+        } else {
+            req.respond(&self.handler, exchange, &metadata.node(), &mut wb)
+                .await?;
+        }
+
+        exchange.send(OpCode::InvokeResponse, wb.as_slice()).await?;
+
+        Ok(())
+    }
+
+    async fn subscribe(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let Some(rx) = self.rx_buffer(exchange).await? else {
+            return Ok(());
+        };
+
+        let Some(mut tx) = self.tx_buffer(exchange).await? else {
+            return Ok(());
+        };
+
+        let req = SubscribeReq::from_tlv(&get_root_node_struct(&rx)?)?;
+        debug!("IM: Subscribe request: {:?}", req);
+
+        let node_id = exchange
+            .with_session(|sess| sess.get_peer_node_id().ok_or(ErrorCode::Invalid.into()))?;
+
+        if !req.keep_subs {
+            self.subscriptions.remove(Some(node_id), None);
+            self.subscriptions_buffers
+                .borrow_mut()
+                .retain(|sb| sb.node_id != node_id);
+
+            info!("All subscriptions for node {node_id:x} removed");
+        }
+
+        let max_int_secs = core::cmp::max(req.max_int_ceil, 40); // Say we need at least 4 secs for potential latencies
+        let min_int_secs = req.min_int_floor;
+
+        if let Some(id) = self.subscriptions.add(node_id, min_int_secs, max_int_secs) {
```

Comment:
nit: to decrease indent maybe:

```rust
let Some(id) = self.subscriptions.add(node_id, min_int_secs, max_int_secs) else {
   return Self::send_status(exchange, IMStatusCode::ResourceExhausted).await;
};
```

or are we trying to not have any `return` somewhere in the middle of functions?

Reply:
I'm honestly undecided on early returns, as to whether they are a good or a bad thing. Most of the time I tried to avoid them, but yes, I've tried using the `let...else` new Rust construct here and there, so why not in subscriptions as wel...

Reply:
Fixed anyway as you suggest.